rust/lance/src/index/vector/utils.rs

```diff
@@ -43,18 +40,13 @@ pub async fn maybe_sample_training_data(
     sample_size_hint: usize,
 ) -> Result<FixedSizeListArray> {
     let num_rows = dataset.count_rows(None).await?;
-    let projection = dataset.schema().project(&[column])?;
     let batch = if num_rows > sample_size_hint {
+        let projection = dataset.schema().project(&[column])?;
         dataset.sample(sample_size_hint, &projection).await?
     } else {
         let mut scanner = dataset.scan();
         scanner.project(&[column])?;
-        let batches = scanner
-            .try_into_stream()
-            .await?
-            .try_collect::<Vec<_>>()
-            .await?;
-        concat_batches(&Arc::new(ArrowSchema::from(&projection)), &batches)?
+        Box::pin(scanner.try_into_batch()).await?
```

Comment:
we are seeing stack overflow caused by large future state on stack.
It looks like some dependency is creating large state and our workload is pushing the stack to limit size.
So here move it to heap